intrinsics.md

```diff
@@ -0,0 +1,183 @@
+# Compiler Intrinsics
+
+* [x] Proposed
+* [ ] Prototype: Not Started
+* [ ] Implementation: Not Started
+* [ ] Specification: Below
+
+## Summary
+
+This proposal provides language constructs that expose low level IL opcodes that cannot currently
+be accessed efficiently, or at all: `ldftn`, `ldvirtftn`, `ldtoken` and `calli`. These low level op 
+codes can be important in high performance code and developers need an efficient way to access 
+them.
+
+## Motivation
+
+The motivations and background for this feature are described in the following isssue (as is a 
+potential implementation of the feature): 
+
+https://github.com/dotnet/csharplang/issues/191
+
+This alternate design proposal comes after reviewing a prototype implementation of the original
+proposal by @msjabby as well as the use throughout a significant code base. This design was done 
+with significant input from @mjsabby, @tmat and @jkotas.
+
+## Detailed Design 
+
+### Allow address of to target methods
+
+Method groups will now be allowed as arguments to an address-of expression. The type of such an 
+expression will be `void*`. 
+
+``` csharp
+class Util { 
+    public static void Log() { } 
+}
+
+// ldftn Util.Log
+void* ptr = &Util.Log; 
+```
+
+Given there is no delegate conversion here the only mechanism for filtering members in the method
+group is by static / instance access. If that cannot distinguish the members then a compile time
+error will occur.
+
+``` csharp
+class Util { 
+    public void Log() { } 
+    public void Log(string p1) { } 
+    public static void Log(int i) { };
+}
+
+unsafe {
+    // Error: Method group Log has more than one applicable candidate.
+    void* ptr1 = &Log; 
+
+    // Okay: only one static member to consider here.
+    void* ptr2 = &Util.Log;
+}
+```
+
+The addressof expression in this context will be implemented in the following manner:
+
+- ldftn: when the method is static 
+- ldvirtftn: when the method is an instance method.
```

Comment:
Would we allow the emission of ldftn if we know that the method is non-virtual?

Reply:
C# always uses `callvirt` instead of `call` and I was trying to mirror that here. I don't have any particular reason beyond that.

Reply:
Having a way to do both `ldvirtftn` and `ldftn` would let library writers determine whether a method has been overridden by a subclass. [This has been used](https://github.com/dotnet/coreclr/blob/95ae6621b1c7f176b518ad819f6eccd469e91ea1/src/System.Private.CoreLib/src/System/IO/Stream.CoreCLR.cs#L12) in performance optimization in the past, but it's unavailable to code that doesn't live in the runtime.

Reply:
BTW: There is nothing that says that the runtime has to return the same pointer from `ldvirtftn` all the time. .NET Framework / Core do actually return different pointers for the same method. 

If the pointer equality is used as an optimization, comparing function pointers will work but reliably.

If the pointer equality is used for correctness, comparing function pointers is not going to work well. (Your HasOverriddenBeginEndRead example is used as both performance optimization but also to achieve correctness.)

Reply:
> There is nothing that says that the runtime has to return the same pointer

We could add public API to compare function pointers for equality, couldn't we? Then this optimization would become available to library developers outside of CoreLib (provided we get a `ldvirtftn`/`ldftn` distinction in this proposal).

Reply:
If we think we can get the CoreFX API approval then it makes sense to be able to distinguish. It won't be a lot more code I can't imagine. @jaredpar what do you think?

Reply:
@mjsabby sorry I'm not quite sure what the suggestion is here. Is it to change which IL opcode we used based on whether the method is virtual vs. non-virtual?

Reply:
> Is it to change which IL opcode we used based on whether the method is virtual vs. non-virtual?

The ask is to provide a way to control whether a `ldftn` or `ldvirtftn` is generated. I somewhat hijacked @mjsabby's thread.

Reply:
Gotcha. I doubt we'd go much further than giving guarantees like the ones already in the proposal. It balances keeping the language pretty simple and gives the developers the lever necessary to get one or the other in the cases where they need it.

On the subject of whether we should use `ldvirtftn` vs `ldftn` on instance methods I'm very open to discussion here. This defaults were chosen just because they matched an existing pattern. Not for a reason specific to this scenario.

Reply:
> On the subject of whether we should use `ldvirtftn` vs `ldftn` on instance methods I'm very open to discussion here

The difference between these two for non-virtual methods would only be about whether we get a `NullReferenceException` for a `null` `this`, right? Since the `this` part is not going to be captured in the returned function pointer, maybe the `&` operator shouldn't care what `this` is for non-virtual methods (and it should use `ldftn`).

It might be convenient to be able to write

```csharp
void* logPtr = &default(Util).Log;
```

Reply:
@MichalStrehovsky 

> The difference between these two for non-virtual methods would only be about whether we get a NullReferenceException for a null this, right? 

For `call` and `callvirt` this also affects binary compatibility. Imagine a method in a utility library changes from non-virtual to virtual and a consuming assembly isn't recompiled. In C# the code in the consuming assembly will automatically begin calling the method as a virtual. 

I'm unsure if `ldftn` and `ldvirtftn` behave in a similar way. Been trying to track that down but haven't found any good dos on it. Assuming they do then it would be a consideration when deciding what code the address-of `&` operator generated.

Reply:
> For `call` and `callvirt` this also affects binary compatibility

Isn't this just a side effect, as opposed from being a conscious design choice? When the compiler knows the `this` is going to non-null, it will happily generate a `call`:

```csharp
    public class Base
    {
        public void Frob()
        {
        }
    }

    class Derived : Base
    {
        public void Call()
        {
            // Emits call
            Frob();
        }

        public static void CallStatic(Derived d)
        {
            // Emits callvirt
            d.Frob();
        }
    }
```

Reply:
> Isn't this just a side effect, as opposed from being a conscious design choice? 

It's hard to say whether in C# 2.0 (where callvirt became the default over call) whether binary compat was a concern. At this point though it's a known quantity of the language that we consider when making changes. It's not an unbreakable change by any means but it's a consideration we make.

Reply:
I think we're over focusing on the callvirt thing here a bit. It's not a big blocking problem so much as it is a box that needs to be checked. I don't suspect there would be any substantial push back from LDM if we make a different decision for how `ldftn` / `ldvirtftn` were chosen.

Reply:
Alright, LGTM.